### PR TITLE
[WIP] feat(Facade): add support for Map.forEach

### PR DESCRIPTION
**Pull request by vicb**
[Review on Reviewable](https://reviewable.io/reviews/angular/ts2dart/215)

### CODE REVIEW

File: lib/facade_converter.ts [hunk 1]
```diff
@@ -244,6 +244,16 @@ export class FacadeConverter extends base.TranspilerBase {
         this.emitCall('remove', c.arguments);
         this.emit('!= null || true ) )');
       },
+      'Map.forEach': (c: ts.CallExpression, context: ts.Expression) => {
+        const callback = <ts.FunctionExpression>c.arguments[0];
+        let params = callback.parameters;
+        if (params.length !== 2) {
+          this.reportError(c, 'Map.forEach callback supports two arguments only');
+        }
+        [params[0], params[1]] = [params[1], params[0]];
```
Comment (vicb) [inline lib/facade_converter.ts]: @mprobst I would prefer creating a new Node instead of mutating an existing one. What do you think ?

**Comment by vicb** [comment]
WIP: when the callback is not an inlined fn (ie references a function) then this should be compiled to 
`m.forEach((k, v) => callback(v, k));

File: lib/facade_converter.ts (hunk 1 above)
Comment (mprobst) [inline lib/facade_converter.ts]: You cannot modify the AST, it must be treated as read only - there's internal state in TypeScript that depends on that. So I'm afraid you'll have to manually handle the callback here, and cannot rely on `visit` doing the right thing.

File: lib/facade_converter.ts [hunk 2]
```diff
@@ -244,6 +244,16 @@ export class FacadeConverter extends base.TranspilerBase {
         this.emitCall('remove', c.arguments);
         this.emit('!= null || true ) )');
       },
+      'Map.forEach': (c: ts.CallExpression, context: ts.Expression) => {
+        const callback = <ts.FunctionExpression>c.arguments[0];
+        let params = callback.parameters;
+        if (params.length !== 2) {
+          this.reportError(c, 'Map.forEach callback supports two arguments only');
```
Comment (mprobst) [inline lib/facade_converter.ts]: nit: `requires two arguments`?

File: lib/facade_converter.ts [hunk 3]
```diff
@@ -244,6 +244,16 @@ export class FacadeConverter extends base.TranspilerBase {
         this.emitCall('remove', c.arguments);
         this.emit('!= null || true ) )');
       },
+      'Map.forEach': (c: ts.CallExpression, context: ts.Expression) => {
+        const callback = <ts.FunctionExpression>c.arguments[0];
```
Comment (mprobst) [inline lib/facade_converter.ts]: What about the case of passing an already bound expression? `map.forEach(myFn)`? I think it's ok to given an error, but we should have a proper message.

Comment (vicb) [inline lib/facade_converter.ts]: I plan to update, see the PR comment

**Pull request closed without merging**